Avoid going back in time in Reflector relist (revived)

jpbetz opened the pull request:
(This is a based on https://github.com/kubernetes/kubernetes/pull/67998. It applies the fix @liggitt suggested in https://github.com/kubernetes/kubernetes/pull/67998#issuecomment-538058258, also addresses component restart and test coverage.)

Fixes the relist part of https://github.com/kubernetes/kubernetes/issues/59848

Update reflectors to relist with resourceVersion="{newest resource version observed}", falling back to resourceVersion="" if an HTTP 410 (Gone) status code is returned by the relist request. This ensures the reflector never "goes back in time" while running.

The pager was also modified to only fallback to a full list if an "Expired" error is returned on page 2 or later. This is to avoid needless reattempts of "too old" requests. See comment in code explaining rationale.

/sig api-machinery
/kind bug
/priority important-soon
@liggitt @wojtek-t @smarterclayton 

```release-note
When the go-client reflector relists, the ResourceVersion list option is set to the reflector's latest synced resource version to ensure the reflector does not "go back in time" and reprocess events older than it has already processed. If the the server responds with an HTTP 410 (Gone) status code response, the relist falls back to using resourceVersion="".
```

wojtek-t committed:
Avoid going back in time in reflector framework

jpbetz committed:
Handle expired errors with RV>0 in pager, don't full list if 1st page is expired

jpbetz commented:
/test pull-kubernetes-integration

liggitt commented on staging/src/k8s.io/client-go/tools/cache/reflector.go:
```diff
@@ -185,10 +188,13 @@ func (r *Reflector) ListAndWatch(stopCh <-chan struct{}) error {
-	// to be served from cache and potentially be delayed relative to
-	// etcd contents. Reflector framework will catch up via Watch() eventually.
-	options := metav1.ListOptions{ResourceVersion: "0"}
+	// Explicitly set resource version to have it list from cache for
```
move this comment to where we set "0" in NewNamedReflector? LastSyncResourceVersion() here isn't necessarily listing from cache

liggitt commented on staging/src/k8s.io/client-go/tools/cache/reflector.go:
```diff
@@ -185,10 +188,13 @@ func (r *Reflector) ListAndWatch(stopCh <-chan struct{}) error {
+	// Explicitly set resource version to have it list from cache for
+	// performance reasons.
+	// It's fine for the returned state to be stale (we will catch up via
+	// Watch() eventually), but can't set "0" to avoid going back in time
```
suggest `It's fine for the returned state to be stale (we will catch up via Watch() eventually), but we need to be at least as new as the last resource version we synced to avoid going back in time.`

the "0" bit is confusing since we don't pass "0" here any more

liggitt commented on staging/src/k8s.io/client-go/tools/cache/reflector.go:
```diff
@@ -114,6 +114,9 @@ func NewNamedReflector(name string, lw ListerWatcher, expectedType interface{},
 		clock:         &clock.RealClock{},
+		// We set lastSyncResourceVersion to "0", because it's the value which
+		// we set as ResourceVersion to the first List() request.
+		lastSyncResourceVersion: "0",
```
there are places that implement `hasSynced` as `return reflector.LastSyncResourceVersion() != ""`

I don't think we should modify the publicly visible behavior of LastSyncResourceVersion to track the resourceVersion we should relist at

jpbetz replied:
Good catch. I've introduced `relistResourceVersion` and reverted `lastSyncResourceVersion` back to initializing to an empty string.

jennybuckley commented:
/cc @yliaog

jpbetz commented:
This is causing real issues with reflector ListAndWatch in TestCRDDeletionCascading. Investigating.

jpbetz committed:
Add HTTP 410 (Gone) status code checks to reflector and relist with RV=''

jpbetz commented:
Code updated to handle HTTP 410 (Gone) status code responses correctly and to start with an initial list using resourceVersion="".  I need to update the tests still.

jpbetz commented on staging/src/k8s.io/client-go/tools/cache/reflector.go:
```diff
@@ -208,19 +211,12 @@ func (r *Reflector) ListAndWatch(stopCh <-chan struct{}) error {
-				// and the list request returned a "Expired" error.
-				options = metav1.ListOptions{ResourceVersion: ""}
-				list, err = pager.List(context.Background(), options)
+			if apierrs.IsResourceExpired(err) || apierrs.IsGone(err) {
```
Reviewers: Worth introducing a new apierrors function to check for all error types that have a  HTTP 410 (Gone) status code? `IsResourceExpired` and `IsGone` check status reason.

jpbetz commented:
Test failures should all be fixed now. This is ready for review.

@smarterclayton, this should solve both of the problems that caused stale reads

jpbetz commented:
/test pull-kubernetes-e2e-gce

liggitt commented on cmd/kubelet/app/server_bootstrap_test.go:
```diff
@@ -324,7 +324,7 @@ func (s *csrSimulator) ServeHTTP(w http.ResponseWriter, req *http.Request) {
 		s.csr = csr
 
-	case req.Method == "GET" && req.URL.Path == "/apis/certificates.k8s.io/v1beta1/certificatesigningrequests" && req.URL.RawQuery == "fieldSelector=metadata.name%3Dtest-csr&limit=500&resourceVersion=0":
+	case req.Method == "GET" && req.URL.Path == "/apis/certificates.k8s.io/v1beta1/certificatesigningrequests" && req.URL.RawQuery == "fieldSelector=metadata.name%3Dtest-csr&limit=500":
```
hmm, does this mean the first request isn't using resourceVersion=0?

jpbetz replied:
Yeah, we're deliberately switching to resourceVersion="" for initial list. I rewrote the description to explain this a bit better, but the basic idea is that we won't want reflectors to go back in time across component restarts, and since we were already already doing a quorum read for resourceVersion="0" (since we skip the watch cache with pagination), let's go ahead and get the strong guarantee that we're getting latest by using resourceVersion="" (which matters for a corner case when the pager falls back to full list).

liggitt replied:
Huh. I **really** didn't think we were paginating initial reflector list calls. If we were, then I guess this won't regress performance, but if we weren't, how do we anticipate this changing performance for node watches of pods (high volume of watchers, across all namespaces, filtered to a small set of total results with spec.nodeName fieldSelector, is probably the worst-case scenario for the API server to pass through to etcd)

liggitt replied:
https://github.com/kubernetes/kubernetes/blob/2bd9643cee5b3b3a5ecbd3af49d09018f0773c77/staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go#L595-L598

The cacher ignores limit requests (which it is permitted to do) when resourceVersion=0.

jpbetz replied:
Good find @liggitt. That makes complete sense. I'll revert the initial list back to resourceVersion="0" and update the title and description of this PR to make it clear that this only prevents the reflector from going back in time for relists, not component restarts.

jpbetz commented:
/retest

jpbetz commented:
cc @oxddr  @wojtek-t @liggitt @mm4tt

redbaron commented on staging/src/k8s.io/client-go/tools/cache/reflector.go:
```diff
@@ -432,3 +443,42 @@ func (r *Reflector) setLastSyncResourceVersion(v string) {
+	if r.lastSyncResourceVersion == "" {
+		// For performance reasons, initial list performed by reflector uses "0" as resource version to allow it to
+		// be served from the watch cache if it is enabled.
+		return "0"
```
Hm, this makes it possible to go back in time on restarts, right? Wouldn't it hurt operators managing statefull resources ?

jpbetz replied:
Yes, this only fixes relist.  Restart has always used resourceVersion=“0” and we’re not fixing that with this PR. We do have plans to fix it, just not with this PR.

jpbetz commented:
Friendly nudge for review

smarterclayton commented on staging/src/k8s.io/client-go/tools/cache/reflector.go:
```diff
@@ -432,3 +443,42 @@ func (r *Reflector) setLastSyncResourceVersion(v string) {
+	r.isLastSyncResourceVersionGone = isExpired
+}
+
+func isGoneError(err error) bool {
```
wait, we don't have an errors helper for this?  Would rather add one.

smarterclayton replied:
I'm really confused why we need the helper vs `apierrs.IsResourceExpired`?

jpbetz replied:
`apierrs.IsResourceExpired` only matches status code `410` where `reason=apierrs.StatusReasonExpired`. But we also have `apierrs.StatusReasonGone` for status code `410`. We must handle both here.

jpbetz replied:
I'll explain in the comment too.

jpbetz replied:
I'm okay with putting a helper in as well. The complication with that is that there is already a `apierrs.IsResourceGone` function that only matches `apierrs.StatusReasonGone` and not `apierrs.StatusReasonExpired`

smarterclayton replied:
Why would the server respond with 410 but not cause expired?  We can have multiple causes, so any place where we expire on server we should set cause expired.

smarterclayton replied:
Put another way, no server code indicating “resource version too old” should return 410 without returning reason expired.  If we have older versions of the server that don’t do that, we should have two checks and a comment on when to remove this check (Ie whenever client lib should stop responding).

jpbetz replied:
Here's where the server returns a 410 using `errors.NewGone`: https://github.com/kubernetes/kubernetes/blob/c5170dfb66749b60a4a1611dc7b4a39e64b33e2e/staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache.go#L471. I'll update that to return `errors.NewResourceExpired` and since that was the last remaining use of `NewGone`, I've marked it as deprecated.

I've swtiched to using two checks and have added a comment explaining why they are both needed and the conditions for removing the one we are deprecating.

smarterclayton commented on staging/src/k8s.io/client-go/tools/cache/reflector.go:
```diff
@@ -432,3 +443,42 @@ func (r *Reflector) setLastSyncResourceVersion(v string) {
+	r.lastSyncResourceVersionMutex.RLock()
+	defer r.lastSyncResourceVersionMutex.RUnlock()
+
+	if r.isLastSyncResourceVersionGone {
```
Only if WatchListPageSize > 0

jpbetz replied:
Not sure I follow. Is the goal to avoid any unpaginated RV="" list request?

If we've already received a 410 Gone, then we can't do a RV="n" request. It would just return Gone again. If we do RV="0", we can get a stale read, which is what we're trying to avoid.

smarterclayton replied:
No, I mean you can create a Reflector that doesn't use paging.  So any logic related to paging has to be constrained to `r.WatchListPageSize > 0`

jpbetz replied:
I think I understand.

I've rewritten the comment explaining the fallback from paginated to full list:

https://github.com/kubernetes/kubernetes/pull/83520/files#diff-df2b34adba86da0485290415620d737aR218

Does that make sense? I believe the logic is correct.

smarterclayton replied:
Yeah, I think so.

jpbetz commented:
rebased

jpbetz commented:
rebased

jpbetz commented:
Friendly nudge to reviewers! I think all feedback has been addressed. I'd like to get this into 1.17 if there are no additional concerns.

smarterclayton commented:
I don't have any additional comments, this is one where I just keep reviewing and reviewing and fearing I missed something.

smarterclayton commented:
/approve

Would like one more person to agree they are happy with this and apply the label.

jpbetz commented:
/retest

wojtek-t commented on staging/src/k8s.io/client-go/tools/cache/reflector.go:
```diff
@@ -432,3 +446,42 @@ func (r *Reflector) setLastSyncResourceVersion(v string) {
+		// Since this reflector makes paginated list requests, and all paginated list requests skip the watch cache
+		// if the lastSyncResourceVersion is expired, we set ResourceVersion="" and list again to re-establish reflector
+		// to the latest available ResourceVersion, using a consistent read from etcd.
+		return ""
```
I hope that with "watch bookmarks" it won't hit us seriously, but it has some potential to break scalability (that said, I agree that we need  it for correctness). I'm just pointing to something that we should look at and if it will appear to happen more than rarely, consider some tuning of watch bookmarks features (or figure out if other things can be done).

liggitt replied:
is the long-term goal to use watch heartbeat on etcd 3.4 to ensure lists at `resourceVersion=0` that are served from the watch cache only return once their data is current with etcd as of the time of the request? would we change this implementation at that point?

jpbetz replied:
I was thinking we'd handle it by making it possible to serve RV="" as a consistent read from watch cache (instead of always hitting etcd for the entire list request), and not guarantee consistency for RV="0".

This should be possible to do. The only needed from etcd to serve consistent reads is which RV is consistent, and then the read can be served from that RV from cache, if the cache has it. etcd has a progress notify operation (progress events are basically bookmarks, but for etcd watch). I've got a sketch of how this might all work, but need to discuss it more broadly before we decide if it's something we'd want to implement. I have a doc I can dust off and circulate.

wojtek-t commented:
/lgtm

I agree with @smarterclayton - I don't see any issues with it, but I also have this fear that we missed something...

wojtek-t commented:
Though - the more soaking time we have, the higher chances to find those issues will be...

liggitt commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache.go:
```diff
@@ -468,7 +468,7 @@ func (w *watchCache) GetAllEventsSinceThreadUnsafe(resourceVersion uint64) ([]*w
 	}
 	if resourceVersion < oldest-1 {
-		return nil, errors.NewGone(fmt.Sprintf("too old resource version: %d (%d)", resourceVersion, oldest-1))
+		return nil, errors.NewResourceExpired(fmt.Sprintf("too old resource version: %d (%d)", resourceVersion, oldest-1))
```
is this a nice-to-have, or does the client implementation depend on this?

jpbetz replied:
This is a nice to have. @smarterclayton had noticed that everywhere else we return resource expired except for there (get,list,watch for watch cache both enabled and disabled). The HTTP Status code is 410 for both responses.

jpbetz replied:
I still check for both on the client side, of course.

liggitt commented:
my only question was about future plans in https://github.com/kubernetes/kubernetes/pull/83520/files#r342871606, lgtm otherwise

unit test failures do look relevant:

> --- FAIL: TestWatch (1.14s)
>     cacher_test.go:370: (called from line 448)
>     cacher_test.go:371: Expected (ERROR): &v1.Status{TypeMeta:v1.TypeMeta{Kind:"", APIVersion:""}, ListMeta:v1.ListMeta{SelfLink:"", ResourceVersion:"", Continue:"", RemainingItemCount:(*int64)(nil)}, Status:"Failure", Message:"", Reason:"Gone", Details:(*v1.StatusDetails)(nil), Code:410}, got: &v1.Status{TypeMeta:v1.TypeMeta{Kind:"", APIVersion:""}, ListMeta:v1.ListMeta{SelfLink:"", ResourceVersion:"", Continue:"", RemainingItemCount:(*int64)(nil)}, Status:"Failure", Message:"too old resource version: 1 (2)", Reason:"Expired", Details:(*v1.StatusDetails)(nil), Code:410}
> 
> --- FAIL: TestEmptyWatchEventCache (6.14s)
>     cacher_test.go:370: (called from line 672)
>     cacher_test.go:371: Expected (ERROR): &v1.Status{TypeMeta:v1.TypeMeta{Kind:"", APIVersion:""}, ListMeta:v1.ListMeta{SelfLink:"", ResourceVersion:"", Continue:"", RemainingItemCount:(*int64)(nil)}, Status:"Failure", Message:"", Reason:"Gone", Details:(*v1.StatusDetails)(nil), Code:410}, got: &v1.Status{TypeMeta:v1.TypeMeta{Kind:"", APIVersion:""}, ListMeta:v1.ListMeta{SelfLink:"", ResourceVersion:"", Continue:"", RemainingItemCount:(*int64)(nil)}, Status:"Failure", Message:"too old resource version: 6 (7)", Reason:"Expired", Details:(*v1.StatusDetails)(nil), Code:410}

/hold
so this doesn't merge on a retest until test failure cause is determined

jpbetz commented:
Test was failing because it expected `Gone` error responses and we've modified the server to always return `ResourceExpired`. I've updated the test to expect `ResourceExpired` instead.

liggitt commented:
Looks like the _output_tests folder accidentally got committed

jpbetz committed:
Fix watch test to expect Expired instead of Gone

jpbetz commented:
> Looks like the _output_tests folder accidentally got committed
> 
> edit: actually, looks like a bunch of random changes got added under vendor

heh, git failed me. That was a from a git add of a single file. The fun part was after that:

```
$ git stash
[1]    133831 segmentation fault  git stash
```

had to back out the commit and nuke my `.git` to restore things.

wojtek-t commented:
/lgtm

liggitt commented:
/hold cancel
